apiserver/src/webhook/mod.rs

```diff
@@ -0,0 +1,316 @@
+mod request;
+mod response;
+
+pub use self::request::{ConversionRequest, Request};
+pub use self::response::{ConversionResponse, ConvertResult, Response};
+
+use models::node::v1::BottlerocketShadow as BottleRocketShadow_v1;
+use models::node::v2::BottlerocketShadow as BottlerocketShadow_v2;
+
+use snafu::{ResultExt, Snafu};
+use std::convert::TryFrom;
+use tracing::instrument;
+
+pub type Result<T> = std::result::Result<T, WebhookConvertError>;
+
+pub fn convert_request_to_response(req: &ConversionRequest) -> ConversionResponse {
+    let request = &req.request;
+    let desired_version = request.desired_api_version.clone();
+
+    match convert_objects(desired_version, request.objects.clone()) {
+        Ok(new_objects) => {
+            let response = Response {
+                uid: request.uid.clone(),
+                result: ConvertResult::default(),
+                converted_objects: Some(new_objects),
+            };
+            ConversionResponse {
+                kind: req.kind.clone(),
+                api_version: req.api_version.clone(),
+                response: response,
+            }
+        }
+        Err(e) => {
+            let fail_result = ConvertResult::create_fail_result(e.to_string());
+            let response = Response {
+                uid: request.uid.clone(),
+                result: fail_result,
+                converted_objects: None,
+            };
+            ConversionResponse {
+                kind: req.kind.clone(),
+                api_version: req.api_version.clone(),
+                response: response,
+            }
+        }
+    }
+}
+
+#[instrument(err)]
+fn convert_objects(
+    desired_version: String,
+    objects: Vec<serde_json::Value>,
+) -> Result<Vec<serde_json::Value>> {
+    let mut new_objects = Vec::new();
+    for old_object in objects.into_iter() {
+        let old_brs_object = BRSObject { object: old_object };
+        let new_brs_object = old_brs_object.chained_convert_object(desired_version.clone())?;
+        new_objects.push(new_brs_object.object);
+    }
+    Ok(new_objects)
+}
+
+struct BRSObject {
+    pub object: serde_json::Value,
+}
+
+impl BRSObject {
+    fn get_version(&self) -> Result<String> {
+        Ok(serde_json::from_value(self.object["apiVersion"].clone())
```

Comment:
Maybe `version` in `BRSObject` should be set once `get_version` is called to prevent double deserialization.